src/rust/ide/src/controller/searcher/suggestion.rs

```diff
@@ -0,0 +1,164 @@
+    pub fn caption(&self) -> &String {
+        match self {
+            Self::Completion(completion) => &completion.name
+        }
+    }
+}
+
+
+
+// ==================
+// === List Entry ===
+// ==================
+
+/// Information how the Suggestion list entry matches the filtering pattern.
+#[allow(missing_docs)]
+#[derive(Clone,Debug,PartialEq)]
+pub enum MatchInfo {
+    DoesNotMatch,
+    Matches {subsequence:fuzzly::Subsequence}
+}
+
+/// The single suggestion list entry.
+#[allow(missing_docs)]
+#[derive(Clone,Debug)]
+pub struct ListEntry {
+    pub match_info : MatchInfo,
+    pub suggestion : Suggestion,
+}
+
+impl ListEntry {
+    /// Update the current match info according to the new filtering pattern.
+    pub fn update_matching_info(&mut self, pattern:impl Str) {
+        let matches     = fuzzly::matches(self.suggestion.caption(),pattern.as_ref());
+        let subsequence = matches.and_option_from(|| {
+            let metric = fuzzly::metric::default();
+            fuzzly::find_best_subsequence(self.suggestion.caption(),pattern,metric)
+        });
+        self.match_info = match subsequence {
+            Some(subsequence) => MatchInfo::Matches {subsequence},
+            None              => MatchInfo::DoesNotMatch,
+        };
+    }
+
+    /// Compare how two entries matches filtering pattern.
+    ///
+    /// The "greater" entry matches better.
+    pub fn compare_match_scores(&self, rhs:&ListEntry) -> std::cmp::Ordering {
+        use MatchInfo::*;
+        use std::cmp::Ordering::*;
+        match (&self.match_info,&rhs.match_info) {
+            (DoesNotMatch             ,DoesNotMatch             ) => Equal,
+            (DoesNotMatch             ,Matches {..}             ) => Less,
+            (Matches {..}             ,DoesNotMatch             ) => Greater,
+            (Matches {subsequence:lhs},Matches {subsequence:rhs}) => lhs.compare_scores(rhs),
+        }
+    }
+}
+
+impl From<Suggestion> for ListEntry {
+    fn from(suggestion:Suggestion) -> Self {
+        let subsequence = default();
+        let match_info  = MatchInfo::Matches {subsequence};
+        ListEntry {match_info,suggestion}
+    }
+}
+
+
+
+// ============
+// === List ===
+// ============
+
+/// Suggestion list.
+///
+/// This structure should be notified about filtering changes. using `update_filtering` function.
+#[derive(Clone,Debug,Default)]
+pub struct List {
+    entries : RefCell<Vec<ListEntry>>
+}
+
+impl List {
+    /// Create new empty list.
+    pub fn new() -> Self {
+        default()
+    }
+
+    /// Create list from suggestions.
+    ///
+    /// The list will assume that the filtering pattern is an empty string.
+    pub fn from_suggestions(suggestions:impl IntoIterator<Item=Suggestion>) -> Self {
+        let entries = suggestions.into_iter().map(ListEntry::from).collect();
+        Self {
+            entries : RefCell::new(entries)
+        }
```

Comment:
You can move `RefCell::new` to firs line and make it 2 lines instead of 4. You can freely ignore this comment here.